### PR TITLE
[MetaSchedule] Logging Interface Unification

### PR DESCRIPTION
**Pull request by zxybazh**
This PR introduces a new interface to unify meta schedule logging from both c++ and python side. It will be easier to control logging level and format content. From now on, only task scheduler level will be printed to screen by default, and task level information will be logged to a file in given folder. We still support any logging configuration from outside, and use `tvm.meta_schedule` as the main logger.

The interface uses logger on the python side while on the c++ side we created a new macro called `TVM_PY_LOG` to work int the following style where `logging_func` is the optional packed function for logging (thanks to #10032 ):
```
TVM_PY_LOG(DEBUG, logging_func)
TVM_PY_LOG(INFO, logging_func)
TVM_PY_LOG(WARNING, logging_func)
TVM_PY_LOG(ERROR, logging_func)
```
The logger would fall back to original c++ logger function `LOG(Level)` if given logging function is not defined.

Recomended setup of logger is as follows:
```
logging.getLogger("tvm.meta_schedule").setLevel(logging.INFO)
# Use logging.DEBUG if debugging
```

And further configuration of task level loggers can be set in `TuneConfig` for features like whether to stream to screen, whether to propagate to main logger, patterns, etc, by passing a Dict to `TuneConfig` as follows:
```
TuneConfig(
    ...,
    logger_config = {
        "loggers": ...,
        "handlers": ...,
        "formatters": ...
    }
    # None if use default setting.
)
```
Default setting is to output to files in `log_dir` folder and the logging structure looks like the following directory:
```
tvm.meta_schedule.tune.task_00_fused_nn_conv2d_add.log
tvm.meta_schedule.tune.task_01_fused_nn_conv2d_add_1.log
tvm.meta_schedule.tune.task_02_fused_nn_conv2d_add_2.log
tvm.meta_schedule.tune.task_03_fused_layout_transform.log
tvm.meta_schedule.tune.task_04_fused_nn_conv2d_add_nn_relu.log
tvm.meta_schedule.tune.task_05_fused_nn_max_pool2d.log
tvm.meta_schedule.tune.task_06_fused_nn_conv2d_add_nn_relu_1.log
tvm.meta_schedule.tune.task_07_fused_nn_conv2d_add_add_nn_relu.log
tvm.meta_schedule.tune.task_08_fused_nn_conv2d_add_nn_relu_2.log
tvm.meta_schedule.tune.task_09_fused_nn_contrib_conv2d_winograd_without_weight_transform_add_nn_relu.log
tvm.meta_schedule.tune.task_10_fused_nn_contrib_conv2d_winograd_without_weight_transform_add_add_nn_relu.log
tvm.meta_schedule.tune.task_11_fused_nn_conv2d_add_nn_relu_3.log
tvm.meta_schedule.tune.task_12_fused_nn_contrib_conv2d_winograd_without_weight_transform_add_nn_relu_1.log
tvm.meta_schedule.tune.task_13_fused_nn_contrib_conv2d_winograd_without_weight_transform_add_add_nn_relu_1.log
tvm.meta_schedule.tune.task_14_fused_nn_conv2d_add_nn_relu_4.log
tvm.meta_schedule.tune.task_15_fused_nn_conv2d_add_nn_relu_5.log
tvm.meta_schedule.tune.task_16_fused_nn_conv2d_add_add_nn_relu_1.log
tvm.meta_schedule.tune.task_17_fused_nn_adaptive_avg_pool2d.log
tvm.meta_schedule.tune.task_18_fused_layout_transform_reshape_squeeze.log
tvm.meta_schedule.tune.task_19_fused_nn_dense_add.log
tvm.meta_schedule.tune.task_scheduler.log
```
Suggestions are welcome!

CC: @junrushao1994 @shingjan @comaniac

### CODE REVIEW

File: include/tvm/meta_schedule/task_scheduler.h [hunk 1]
```diff
@@ -96,6 +98,7 @@ class TaskSchedulerNode : public runtime::Object {
     v->Visit("cost_model", &cost_model);
     v->Visit("measure_callbacks", &measure_callbacks);
     v->Visit("num_trials_already", &num_trials_already);
+    // v->Visit("logging_func", &logging_func);
```
Comment (junrushao1994) [inline include/tvm/meta_schedule/task_scheduler.h:101]: explicitly say it's skipped 

```suggestion
    // `logging_func` is not visited
```

File: python/tvm/meta_schedule/apply_history_best.py [hunk 2]
```diff
@@ -25,6 +26,8 @@
 from . import _ffi_api
 from .database import Database
 
+logger = logging.getLogger("tvm.meta_schedule")  # pylint: disable=invalid-name
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/apply_history_best.py:29]:
```suggestion
logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
```

File: python/tvm/meta_schedule/apply_history_best.py [hunk 3]
```diff
@@ -34,15 +37,17 @@ class ApplyHistoryBest(Object):
     ----------
     database : Database
         The database to be queried from
+    logger : logging.Logger
+        The logger to be used
     """
 
     database: Database
+    logger: logging.Logger
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/apply_history_best.py:45]: remove this given it's not used

File: python/tvm/meta_schedule/cost_model/xgb_model.py [hunk 4]
```diff
@@ -41,7 +41,7 @@
     from ..tune_context import TuneContext
 
 
-logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+logger = logging.getLogger("tvm.meta_schedule")  # pylint: disable=invalid-name
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/cost_model/xgb_model.py:44]: revert this

File: python/tvm/meta_schedule/runner/local_runner.py [hunk 5]
```diff
@@ -33,7 +33,8 @@
     run_evaluator_common,
 )
 
-logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+logger = logging.getLogger("tvm.meta_schedule")  # pylint: disable=invalid-name
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/runner/local_runner.py:37]: ditto

File: python/tvm/meta_schedule/runner/rpc_runner.py [hunk 6]
```diff
@@ -39,8 +39,8 @@
     run_evaluator_common,
 )
 
-logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
 
+logger = logging.getLogger("tvm.meta_schedule")  # pylint: disable=invalid-name
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/runner/rpc_runner.py:43]: ditto

File: python/tvm/meta_schedule/task_scheduler/gradient_based.py [hunk 7]
```diff
@@ -15,9 +15,11 @@
 # specific language governing permissions and limitations
 # under the License.
 """Gradient Based Task Scheduler"""
+import logging
 from typing import TYPE_CHECKING, List, Optional
 
 from tvm._ffi import register_object
+from tvm.meta_schedule.utils import make_logging_func
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/task_scheduler/gradient_based.py:22]: use relative import:

```suggestion
from ..utils import import make_logging_func
```

File: python/tvm/meta_schedule/task_scheduler/round_robin.py [hunk 8]
```diff
@@ -16,10 +16,12 @@
 # under the License.
 """Round Robin Task Scheduler"""
 
+import logging
 from typing import TYPE_CHECKING, List, Optional
 
 from tvm._ffi import register_object
 from tvm.meta_schedule.measure_callback.measure_callback import MeasureCallback
+from tvm.meta_schedule.utils import make_logging_func
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/task_scheduler/round_robin.py:24]: ditto

File: python/tvm/meta_schedule/task_scheduler/gradient_based.py [hunk 9]
```diff
@@ -87,6 +92,7 @@ def __init__(
             max_trials,
             cost_model,
             measure_callbacks,
+            make_logging_func(logger),
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/task_scheduler/gradient_based.py:95]: to be clear, the logger on this file should be configured as 

```
logger = logging.getLogger(__name__)
```

File: python/tvm/meta_schedule/testing/relay_workload.py [hunk 10]
```diff
@@ -54,6 +54,8 @@ def _get_network(
         "resnet3d_18",
         "vgg_16",
     ]:
+        # pylint: disable=import-outside-toplevel
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/testing/relay_workload.py:57]: no need to tweak this

**Comment by zxybazh** [comment]
Thanks for the careful review, I've fixed all the logger usage and made sure it follows the current recommended practice. Ready for review.

File: python/tvm/meta_schedule/task_scheduler/task_scheduler.py [hunk 11]
```diff
@@ -16,9 +16,11 @@
 # under the License.
 """Auto-tuning Task Scheduler"""
 
+import logging
 from typing import Callable, List, Optional
 
 from tvm._ffi import register_object
+from tvm.meta_schedule.utils import make_logging_func
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/task_scheduler/task_scheduler.py:23]: use relative import inside a subpackage

File: src/meta_schedule/utils.h [hunk 12]
```diff
@@ -53,7 +53,49 @@
...
 
+#define TVM_PY_LOG(logging_level, logging_func) \
+  PyLogMessage(__FILE__, __LINE__, logging_func, PyLogMessage::Level::logging_level).stream()
 namespace tvm {
+
+/*!
+ * \brief Class to accumulate an log message on the python side. Do not use directly, instead use
+ * TVM_PY_LOG(INFO), TVM_PY_LOG(WARNING), TVM_PY_ERROR(INFO).
+ */
+class PyLogMessage {
```
Comment (junrushao1994) [inline src/meta_schedule/utils.h:64]: Move this class into tvm::meta_schedule

File: src/meta_schedule/utils.h [hunk 13]
```diff
@@ -53,7 +53,49 @@
 #include "../tir/schedule/primitive.h"
 #include "../tir/schedule/utils.h"
 
+#define TVM_PY_LOG(logging_level, logging_func) \
+  PyLogMessage(__FILE__, __LINE__, logging_func, PyLogMessage::Level::logging_level).stream()
```
Comment (junrushao1994) [inline src/meta_schedule/utils.h:57]:
```suggestion
  ::tvm::meta_schedule::PyLogMessage(__FILE__, __LINE__, logging_func, PyLogMessage::Level::logging_level).stream()
```

File: python/tvm/meta_schedule/utils.py [hunk 14]
```diff
@@ -368,3 +372,30 @@ def autotvm_silencer():
         yield
     finally:
         autotvm.GLOBAL_SCOPE.silent = silent
+
+
+def make_logging_func(logger: logging.Logger):
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/utils.py:377]:
```suggestion
def make_logging_func(logger: logging.Logger) -> Optional[Callable]:
```

File: python/tvm/meta_schedule/utils.py [hunk 15]
```diff
@@ -368,3 +372,30 @@ def autotvm_silencer():
...
+    """Get the logging function.
+    Parameters
+    ----------
+    logger : logging.Logger
+        The logger instance.
+    Returns
+    -------
+    result : Callable
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/utils.py:385]:
```suggestion
    result : Optional[Callable]
```

File: python/tvm/meta_schedule/tune.py [hunk 16]
```diff
@@ -17,7 +17,8 @@
 """User-facing Tuning API"""
 # pylint: disable=import-outside-toplevel
 import logging
-import os.path
+from pathlib import Path
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/tune.py:20]: Pathlib is definitely a good library, but perhaps it's less plausible if we mix `os.path` and `Pathlib`. In our particular case, shall we just do:

```python
os.makedirs(path, exist_ok=False)
```

File: python/tvm/meta_schedule/tune.py [hunk 17]
```diff
@@ -493,6 +544,11 @@ def tune_extracted_tasks(
                 postprocs=Parse._postproc(postprocs, task.target),
                 mutator_probs=Parse._mutator_probs(mutator_probs, task.target),
                 task_name=task.task_name,
+                logger=Parse._logger(
+                    name="tvm.meta_schedule."
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/tune.py:548]:
```suggestion
                    name=__name__ + "."
```

File: python/tvm/meta_schedule/tune.py (hunk 16 above)
Comment (zxybazh) [reply python/tvm/meta_schedule/tune.py:20]: I think it might be acceptable to use `exist_ok = True` in this case cause the logs folder could be reused.

File: python/tvm/meta_schedule/testing/tune_relay_meta_schedule.py [hunk 18]
```diff
@@ -90,8 +90,10 @@ def _parse_args():
     return parsed
 
 
-logging.basicConfig()
-logging.getLogger("tvm.meta_schedule").setLevel(logging.DEBUG)
+logging.basicConfig(
+    format="%(asctime)s.%(msecs)03d %(levelname)s %(message)s", datefmt="%Y-%m-%d %H:%M:%S"
+)
+logging.getLogger("tvm.meta_schedule").setLevel(logging.INFO)
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/testing/tune_relay_meta_schedule.py:96]: Shall we also use this in `tune_te_meta_schedule.py`?

File: src/meta_schedule/schedule_rule/cross_thread_reduction.cc [hunk 19]
```diff
@@ -32,12 +32,14 @@ class CrossThreadReductionNode : public ScheduleRuleNode {
     Optional<Integer> opt_warp_size = target->GetAttr<Integer>("thread_warp_size");
 
     if (!opt_max_threads_per_block.defined()) {
-      LOG(WARNING) << "Target does not have attribute \"max_threads_per_block\", therefore the "
-                      "rule CrossThreadReduction will not be applied";
+      TVM_PY_LOG(WARNING, context->logging_func)
```
Comment (junrushao1994) [inline src/meta_schedule/schedule_rule/cross_thread_reduction.cc:35]: I was thinking if it's better to make it a LOG(FATAL) instead, but it's do that in a separate PR

File: src/meta_schedule/utils.h [hunk 20]
```diff
@@ -355,6 +403,8 @@ struct ThreadedTraceApply {
   int n_;
   /*! \brief The pointer to the list of postprocessor items. */
   Item* items_;
+  /*! \brief The logging function to use. */
+  PackedFunc logging_func;
```
Comment (junrushao1994) [inline src/meta_schedule/utils.h:407]: We don't need this field in this class :-)

**Comment by junrushao1994** [comment]
Any updates?

**Review by Hzfengsy — APPROVED** [review]

**Review by junrushao1994 — APPROVED** [review]
LGTM! Please try it out for the last time and confirm it prints the dates properly, and then feel free to merge it yourself

**Comment by zxybazh** [comment]
Thanks @junrushao1994 @Hzfengsy for reviewing!
